src/MagnumPlugins/CgltfImporter/Test/CgltfImporterTest.cpp

```diff
@@ -0,0 +1,4435 @@
+            }), TestSuite::Compare::Container);
+    } else CORRADE_VERIFY(!mesh->hasAttribute(MeshAttribute::ObjectId));
+}
+
+void CgltfImporterTest::meshOutOfBounds() {
+    auto&& data = MeshOutOfBoundsData[testCaseInstanceId()];
+    setTestCaseDescription(data.name);
+
+    Containers::Pointer<AbstractImporter> importer = _manager.instantiate("CgltfImporter");
+
+    std::ostringstream out;
+    Error redirectError{&out};
+    CORRADE_VERIFY(!importer->openFile(Utility::Directory::join(TINYGLTFIMPORTER_TEST_DIR, data.file)));
+    CORRADE_COMPARE(out.str(), "Trade::CgltfImporter::openData(): error opening file: invalid glTF\n");
+}
+
+void CgltfImporterTest::meshInvalid() {
+    auto&& data = MeshInvalidData[testCaseInstanceId()];
+    setTestCaseDescription(data.name);
+
+    Containers::Pointer<AbstractImporter> importer = _manager.instantiate("CgltfImporter");
+
+    CORRADE_VERIFY(importer->openFile(Utility::Directory::join(TINYGLTFIMPORTER_TEST_DIR,
+        "mesh-invalid.gltf")));
+
+    std::ostringstream out;
+    Error redirectError{&out};
+    CORRADE_VERIFY(!importer->mesh(data.name));
+    CORRADE_COMPARE(out.str(), Utility::formatString("Trade::CgltfImporter::mesh(): {}\n", data.message));
+}
+
+void CgltfImporterTest::meshInvalidSkinAttributes() {
+    auto&& data = MeshInvalidSkinAttributesData[testCaseInstanceId()];
+    setTestCaseDescription(data.name);
+
+    Containers::Pointer<AbstractImporter> importer = _manager.instantiate("CgltfImporter");
+
+    CORRADE_VERIFY(importer->openFile(Utility::Directory::join(CGLTFIMPORTER_TEST_DIR,
+        "mesh-invalid-skin-attributes.gltf")));
+
+    std::ostringstream out;
+    Error redirectError{&out};
+    CORRADE_VERIFY(!importer->mesh(data.name));
+    CORRADE_COMPARE(out.str(), Utility::formatString("Trade::CgltfImporter::mesh(): {}\n", data.message));
+}
+
+void CgltfImporterTest::meshInvalidTypes() {
+    auto&& data = MeshInvalidTypesData[testCaseInstanceId()];
+    setTestCaseDescription(data.name);
+
+    Containers::Pointer<AbstractImporter> importer = _manager.instantiate("CgltfImporter");
+
+    CORRADE_VERIFY(importer->openFile(Utility::Directory::join(CGLTFIMPORTER_TEST_DIR, data.file)));
+
+    std::ostringstream out;
+    Error redirectError{&out};
+    CORRADE_VERIFY(!importer->mesh(data.name));
+    CORRADE_COMPARE(out.str(), Utility::formatString("Trade::CgltfImporter::mesh(): {}\n", data.message));
+}
+
+void CgltfImporterTest::meshInvalidIndexComponentType() {
+    Containers::Pointer<AbstractImporter> importer = _manager.instantiate("CgltfImporter");
+
+    /* cgltf_validate checks index accessor types, but only if the mesh has any
+       attributes. The test case in mesh-invalid.gltf has no attributes, so it
+       only tests our own check. This one tests the result of cgltf_validate.
+       This is all a bit shitty since the check is now duplicated and occurs in
+       different functions (openData vs. mesh) and with different error text. */
```

Comment:
Ah well. I was just wondering why there's the additional `mesh-invalid-*.gltf` files. Together with my other comment about underwhelming error messages, would it make sense to reimplement `cgltf_validate()` on our side to have the handling more consistent? Basically we'd have to always look at what it does and then add the missing checks on our side (especially for extensions cgltf doesn't support yet), but instead if we'd be handling everything then it's much easier to ensure all cases are handled properly.

Reply:
Doing our own validation is probably the better option, indeed. Saves us from silly edge cases like above, adds useful (and consistent) error messages and moves some checks to only run when really needed. I initially dismissed it because of code duplication, but I think the advantages overweigh :eyes: